### PR TITLE
Iox #415 add event parameter to find service

### LINKED ISSUE #415: service registry as a built-in topic
## Brief feature description

The discovery requests from user side are currently transferred via the message queue connection. This has some drawbacks

- The message queue is not only used for registration and creation of resources during startup but also during runtime
- If an application polls the discovery information with a high frequency the message queue becomes a bottleneck 
- We have a message size limitation on the message queue or unix domain socket, i.e. we need some transport protocol to split the response to several messages if there are a lot services. Currently it is a limitation as we do not have such a protocol 
- The `findService()` is very close to AUTOSAR ara::com and maybe cannot be used in a reasonable way in other frameworks 
- If someone wants to use ara::com service discovery callbacks, having only a polling based interface is a big drawback 

## Detailed information

The proposal is to have a service registry that provides it's content as a built-in topic, similar to the introspection topics. Then a subscriber could be created to get the information either in a polling way or event-driven by using the subscriber in combination with a waitset. The service registry publisher would use a history with size 1. Then every subscriber gets the current state on subscribe if desired. By providing the full CaPro ID information and not only service and instance IDs, this built-in topic could also be used for non AUTOSAR use cases. With this approach having a large service registry is fine as we don't have to send the content over the message queue and it's distributed with iceoryx' finest true zero-copy. I guess we can then also get rid of `getServiceRegistryChangeCounter()` which smells like workaround

## Todo
- [x] Create design decision document
- [x] Introduce poor mans `cxx::map` (Will be done in #829)
- [x] Refactor `ServiceDescription`
  - [x] Remove the two argument c'tor in `ServiceDescription`
  - [x] Remove three integers in `ServiceDescription`
  - [x] ~~Forbid creating Pub/Sub with `ServiceDescription(AnyServiceString, AnyInstanceString, "foo")` and forbid adding such a invalid `ServiceDescription` to the registry~~
  - [x] ~~Add [ServiceDescriptionBuilder](https://github.com/eclipse-iceoryx/iceoryx/pull/1047#discussion_r792652416)~~ Moved to #1141
- [x] Refactor `ServiceRegistry` to use the whole `ServiceDescription`
- [x] Error handling in `ServiceDescription(const cxx::Serialization& serial)` like `client_options.cpp`
- [x] Create example for new user-facing `DiscoveryInfo` class and move `getServiceRegistryChangeCounter`, `findService`, ~~`offerService` and `stopOfferService`~~ to this class 
  - [x] Extend `icediscovery`
  - [x] Write `README.md` for `icediscovery`
- ~~[ ] Introduce new [`StatusPorts`](https://github.com/eclipse-iceoryx/iceoryx/discussions/818)~~ Moved to #982
- [ ] ~~Built-in topic for service registry based on`BroadcastPorts`~~ Moved to #1135
   - [ ] ~~User facing API~~
- ~~[ ] Replace `InterfacePorts` with `StatusPorts`~~ Moved to #982
- [ ] ~~Mapping to Autosar Adaptive and APIs to work on an Autosar service level~~ Moved to #1103
- ~~[ ] Usage of #830~~
  - ~~[ ] Adaption of `ServiceRegistry::add()` (remove duplicate search)~~ (not needed, we basically have a multiset with counters)
- [x] `ServiceRegistry::remove()` should use same algorithm as `find()` to locate the entry (and then remove it)

Discussions:
- [x] Consider `ServiceRegistry::find("Foo", Wildcard, [] (auto& serviceDescription) { })` [API](https://github.com/eclipse-iceoryx/iceoryx/pull/860#discussion_r693374036)
- [x] Can `ServiceDescription::isValid()` be avoided?
- [x] When `ManyToManyPolicy` is used, would a user be interested in the number of offers? This is stored in the `ServiceRegistry` as `ServiceDescriptionEntry::referenceCounter` but currently not passed up to the user API.
- [ ] ~~Does the `ServiceRegistry` belong to `roudi` or `runtime`? Somehow, it sits in-between, currently it is in `source/roudi/service_registry.cpp`~~ Moved to #1200

Stretch goals:
- [ ] ~~Naming of the CaPro ID members, consider [comment](https://github.com/eclipse-iceoryx/iceoryx/pull/481/files#r553892569) from @ithier~~  Moved to #1200
- [ ] ~~Storage of IDs in service registry and ports (short IDs that are references to the string-based IDs, dictionary)~~ Moved to #1200
- [ ] ~~Decision on whether to remove wildcard in Cyclone DDS gateway~~ Moved to #1200
- [x] Check if `FindServiceError::UNABLE_TO_WRITE_TO_ROUDI_CHANNEL` can be removed
- [x] Check if `CaproMessageSubType` can be removed

### PR DESCRIPTION
**Pull request by FerdinandSpitzschnueffler**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

## Notes for Reviewer

This PR adds an event parameter to the find service methods. The event search inside `ServiceRegistry` is probably not implemented in the most efficient way, but will be replaced soon when the [prefix tree](https://github.com/eclipse-iceoryx/iceoryx/pull/1057) is integrated.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Partially closes #415

### CODE REVIEW

**Review by elfenpiff — DISMISSED** [review]

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 1]
```diff
@@ -152,14 +155,29 @@ void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
...
+        {
+            if (iterator->serviceDescription.getEventIDString() == event.value())
+            {
+                searchResult.push_back(*iterator);
+            }
+        }
+    }
+    else
+    {
+        searchResult = serviceInstanceResult;
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/roudi/service_registry.cpp:180]: This logically correct but costly (as we cannot move the result). We will probably change it soon anyway. Ideally we would fill the search result once in the way it is needed (in this branch the `serviceInstanceResult` is already the search result.

As it is one way would probably need to check earlier (e.g. in each case) whether we have an `event` wildcard.

But this would make the code less concise and it is not optimized for efficiency now anyway. I do not see what can be done to remove this inefficiency in a simple way.

File: iceoryx_posh/source/runtime/service_discovery.cpp [hunk 2]
```diff
@@ -39,11 +42,16 @@ ServiceDiscovery::findService(const cxx::optional<capro::IdString_t>& service,
     {
         instanceString = instance.value();
     }
+    if (!isEventWildcard)
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/runtime/service_discovery.cpp:45]: Can probably now use `if(event)` and get rid of the helper `bool isEventWildcard` (introduced before `optional`) and similar `bool` values. Will also get rid of redundat negation etc.

File: iceoryx_posh/source/runtime/service_discovery.cpp [hunk 3]
```diff
@@ -23,13 +23,16 @@ namespace runtime
...
 {
     /// @todo #415 remove the string mapping, once the find call is done via shared memory
     capro::IdString_t serviceString;
     capro::IdString_t instanceString;
+    capro::IdString_t eventString;
     bool isServiceWildcard = !service;
     bool isInstanceWildcard = !instance;
+    bool isEventWildcard = !event;
 
     if (!isServiceWildcard)
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/runtime/service_discovery.cpp:37]: Could use `if(service)`, see below.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp [hunk 4]
```diff
@@ -177,18 +177,29 @@ TEST_F(ServiceRegistry_test, SingleServiceDescriptionCanBeFoundWithWildcardSearc
...
     auto result = sut.add(ServiceDescription("Foo", "Bar", "Baz"));
     ASSERT_FALSE(result.has_error());
-    sut.find(searchResults, iox::capro::Wildcard, iox::capro::Wildcard);
+    sut.find(searchResults, iox::capro::Wildcard, iox::capro::Wildcard, iox::capro::Wildcard);
 
     EXPECT_THAT(searchResults.size(), Eq(1));
     EXPECT_THAT(searchResults[0].serviceDescription, Eq(ServiceDescription("Foo", "Bar", "Baz")));
 }
 
+TEST_F(ServiceRegistry_test, SingleServiceDescriptionCanBeFoundWithEventName)
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:186]: I think it makes sense to have another test where we search for an event that does not exist (and expect an empty result).

Probably instance and service as well (is this done?), but not in this PR as it is concerned with events.

It is probably necessary after we e.g. change the search algorithm in the `ServiceRegistry` we probably need to evaluate whether the tests actually cover all relevant cases.

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp [hunk 5]
```diff
@@ -222,28 +230,79 @@ TEST_F(ServiceDiscovery_test, FindServiceWithInstanceWildcardReturnsAllMatchingS
...
+    const IdString_t SERVICE = "service";
+    const IdString_t EVENT = "event";
+    const iox::capro::ServiceDescription SERVICE_DESCRIPTION1(SERVICE, "instance1", EVENT);
+    iox::popo::UntypedPublisher publisher_sd1(SERVICE_DESCRIPTION1);
+    const iox::capro::ServiceDescription SERVICE_DESCRIPTION2("another_service", "instance2", EVENT);
+    iox::popo::UntypedPublisher publisher_sd2(SERVICE_DESCRIPTION2);
+    const iox::capro::ServiceDescription SERVICE_DESCRIPTION3(SERVICE, "instance3", EVENT);
+    iox::popo::UntypedPublisher publisher_sd3(SERVICE_DESCRIPTION3);
+    const iox::capro::ServiceDescription SERVICE_DESCRIPTION4(SERVICE, "instance4", "another_event");
+    iox::popo::UntypedPublisher publisher_sd4(SERVICE_DESCRIPTION4);
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/integrationtests/test_service_discovery.cpp:299]: I understand the structure of the tests but somehow it looks like the naming of variables and strings makes it a little harder to follow then it needs to be.

Maybe we could use trailing numbers in the strings consistently and rename `publisher_sdx` into `publisher_x`.
I think this PR mainly used the existing scheme and this is fine.

The tests added make sense, but I have a hard time to say whether we are not missing something.

We could also create a new issue (stretch goal) to refactor `ServiceDiscovery` tests as needed (including analysis whether there are missing tests and adding missing tests).

**Review by MatthiasKillat — CHANGES_REQUESTED** [review]

File: iceoryx_posh/source/roudi/service_registry.cpp (hunk 1 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_posh/source/roudi/service_registry.cpp:180]: As I stated in the `Notes for Reviewer`, I know it's quite inefficient. Since we probably will change the underlying data structure soon, is it okay for you to leave it as it is @MatthiasKillat ?

File: iceoryx_posh/source/runtime/service_discovery.cpp (hunk 2 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_posh/source/runtime/service_discovery.cpp:45]: Correct, but I kept `is*Wildcard` booleans to make the intention clearer in line 52ff. But no hard feelings about this.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp (hunk 4 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:186]: Done for service and instance. I'll add another test where the service description searched for only differs in the event.

File: iceoryx_posh/source/roudi/service_registry.cpp (hunk 1 above)
Comment (MatthiasKillat) [reply iceoryx_posh/source/roudi/service_registry.cpp:180]: Sure. And I missed the notes somehow.

File: iceoryx_posh/source/runtime/service_discovery.cpp (hunk 2 above)
Comment (MatthiasKillat) [reply iceoryx_posh/source/runtime/service_discovery.cpp:45]: I would prefer to get rid of them because for me `if(event)` reads nicely as "if the event exists" (i.e. is no wildcard). Gets rid of unnecessary definitions and saves (maybe optimized) negations.

File: CHANGELOG.md [hunk 6]
```diff
@@ -236,7 +237,7 @@ poshRuntime.stopOfferService(myServiceDescription);
 poshRuntime.findService({"ServiceA", iox::capro::AnyInstanceString});
 
 // after
-serviceDiscovery.findService("ServiceA", Wildcard);
+serviceDiscovery.findService("ServiceA", Wildcard, Wildcard);
```
Comment (elfenpiff) [inline CHANGELOG.md:240]: Since the include has also changed could you please add the new include directive needed to use the service discovery - only if it is required.

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 7]
```diff
@@ -96,19 +97,21 @@ void ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription
...
 void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const cxx::optional<capro::IdString_t>& service,
-                           const cxx::optional<capro::IdString_t>& instance) const noexcept
+                           const cxx::optional<capro::IdString_t>& instance,
+                           const cxx::optional<capro::IdString_t>& event) const noexcept
 {
-    cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
-
+    ServiceDescriptionVector_t serviceInstanceResult;
     // Find (K1, K2)
```
Comment (elfenpiff) [inline iceoryx_posh/source/roudi/service_registry.cpp:110]: Now that event is added could we please remove those runtime big O notation comments.

**Review by elfenpiff — COMMENTED** [review]
Could you please create an issue (or add to some existing issue) that we have to refactor the `test_service_discovery.cpp`. This is pretty hard to understand and not our standard - but also not part of this PR.

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp (hunk 5 above)
Comment (elfenpiff) [reply iceoryx_posh/test/integrationtests/test_service_discovery.cpp:299]: I would not invest any more time into those service discovery tests here. Leave them be and another poor soul has to refactor them completely sooner or later.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp [hunk 8]
```diff
@@ -177,18 +177,29 @@ TEST_F(ServiceRegistry_test, SingleServiceDescriptionCanBeFoundWithWildcardSearc
...
+{
+    ::testing::Test::RecordProperty("TEST_ID", "6df8fd7d-e4d1-4c51-8ad7-2fbe82e4ed09");
+    iox::capro::ServiceDescription service("a", "b", "c");
+    ASSERT_FALSE(sut.add(service).has_error());
+    sut.find(searchResults, iox::capro::Wildcard, iox::capro::Wildcard, iox::capro::IdString_t("c"));
+
+    EXPECT_THAT(searchResults.size(), Eq(1));
```
Comment (mossmaurice) [inline iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:193]:
```suggestion
    ASSERT_THAT(searchResults.size(), Eq(1));
```

**Review by mossmaurice — COMMENTED** [review]
LGTM in general!

File: iceoryx_posh/source/runtime/service_discovery.cpp (hunk 2 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_posh/source/runtime/service_discovery.cpp:45]: 3:1 for keeping `is*Wildcard`.

File: iceoryx_posh/source/roudi/service_registry.cpp (hunk 7 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_posh/source/roudi/service_registry.cpp:110]: I didn't check the O notation for correctness, but the branches where it's noted didn't change. Therefore, no updates needed now. Should be updated (or removed) when the prefix tree is integrated.

File: iceoryx_posh/test/integrationtests/test_service_discovery.cpp (hunk 5 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_posh/test/integrationtests/test_service_discovery.cpp:299]: Created #1065

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_posh/source/roudi/service_registry.cpp (hunk 7 above)
Comment (MatthiasKillat) [reply iceoryx_posh/source/roudi/service_registry.cpp:110]: I will update this/remove as needed.
